test/fuse/linux/read_test.cc

```diff
@@ -0,0 +1,236 @@
+  SetServerInodeLookup(test_file_, S_IFREG | S_IRWXU | S_IRWXG | S_IRWXO);
+  const int open_flag = O_RDWR;
+  struct fuse_out_header out_header = {
+      .len = sizeof(struct fuse_out_header) + sizeof(struct fuse_open_out),
+  };
+  struct fuse_open_out out_payload = {
+      .fh = 1,
+      .open_flags = open_flag,
+  };
+  auto iov_out_open = FuseGenerateIovecs(out_header, out_payload);
+  SetServerResponse(FUSE_OPEN, iov_out_open);
+  FileDescriptor fd =
+      ASSERT_NO_ERRNO_AND_VALUE(Open(test_file_path.c_str(), open_flag));
+
+  // Can replace with SkipServerActualRequest() after mergerd.
+  struct fuse_in_header in_header_open;
+  struct fuse_open_in in_payload_open;
+  auto iov_in_open = FuseGenerateIovecs(in_header_open, in_payload_open);
+  GetServerActualRequest(iov_in_open);
+
+  EXPECT_EQ(in_header_open.len,
+            sizeof(in_header_open) + sizeof(in_payload_open));
+  EXPECT_EQ(in_header_open.opcode, FUSE_OPEN);
+
+  // Prepare for the read.
+  const int n_read = 5;
+  std::string data = std::string(n_read, 'a');
+  struct fuse_out_header read_out_header = {
+      .len = static_cast<uint32_t>(sizeof(struct fuse_out_header) +
+                                   data.size() + 1),
+  };
+  auto iov_out_read = FuseGenerateIovecs(read_out_header, data);
+  SetServerResponse(FUSE_READ, iov_out_read);
+
+  // Read the whole "file".
+  char buf[n_read + 1];
+  EXPECT_THAT(read(fd.get(), buf, n_read), SyscallSucceedsWithValue(n_read));
+  buf[n_read] = '\0';
+
+  // Check the read request.
+  struct fuse_in_header in_header;
+  struct fuse_read_in in_payload;
+  auto iov_in = FuseGenerateIovecs(in_header, in_payload);
+  GetServerActualRequest(iov_in);
+
+  EXPECT_EQ(in_header.len, sizeof(in_header) + sizeof(in_payload));
+  EXPECT_EQ(in_header.opcode, FUSE_READ);
+  EXPECT_EQ(in_payload.offset, 0);
+  EXPECT_STREQ(buf, data.c_str());
+}
+
+TEST_F(ReadTest, ReadPartial) {
+  const std::string test_file_path =
+      JoinPath(mount_point_.path().c_str(), test_file_);
+
+  // Open to get a fd.
+  SetServerInodeLookup(test_file_, S_IFREG | S_IRWXU | S_IRWXG | S_IRWXO);
+  const int open_flag = O_RDWR;
+  struct fuse_out_header out_header = {
+      .len = sizeof(struct fuse_out_header) + sizeof(struct fuse_open_out),
+  };
+  struct fuse_open_out out_payload = {
+      .fh = 1,
+      .open_flags = open_flag,
+  };
+  auto iov_out_open = FuseGenerateIovecs(out_header, out_payload);
+  SetServerResponse(FUSE_OPEN, iov_out_open);
+  FileDescriptor fd =
+      ASSERT_NO_ERRNO_AND_VALUE(Open(test_file_path.c_str(), open_flag));
+
+  // Can replace with SkipServerActualRequest() after mergerd.
+  struct fuse_in_header in_header_open;
+  struct fuse_open_in in_payload_open;
+  auto iov_in_open = FuseGenerateIovecs(in_header_open, in_payload_open);
+  GetServerActualRequest(iov_in_open);
+
+  EXPECT_EQ(in_header_open.len,
+            sizeof(in_header_open) + sizeof(in_payload_open));
+  EXPECT_EQ(in_header_open.opcode, FUSE_OPEN);
```

Comment:
Looks like the same across tests. Want to declare as a function in class `ReadTest`?

Reply:
FUSE_READ is not finished as a complete operation for demo purpose. Will add the missing ones (might need to change the fuse_base.cc) afterwards.